File: src/redactor.ts
import type { GmailMessage } from './gmail/fake.js';
import type { RedactionRule } from './rules.js';

/**
 * Deterministic, non-LLM redactor. The base `redact` function strips the
 * message body and reduces `from` to its domain. `applyRedactionRules`
 * applies additional field-level and regex rules from principles.md.
 */
export interface RedactedMessage {
  id: string;
  fromDomain: string;
  subject: string;
}

export function redact(message: GmailMessage): RedactedMessage {
  const at = message.from.lastIndexOf('@');
  const fromDomain = at >= 0 ? message.from.slice(at + 1) : message.from;
  return { id: message.id, fromDomain, subject: message.subject };
}

/**
 * Apply rule-driven redaction on top of the base redaction.
 * - Rules with no pattern replace the entire field with [REDACTED].
 * - Rules with a pattern replace all regex matches with [REDACTED].
 * Returns a new object; does not mutate the input.
 */
export function applyRedactionRules(
  message: RedactedMessage,
  rules: RedactionRule[],
): RedactedMessage {
  if (rules.length === 0) return message;

  const result: Record<string, unknown> = { ...message };

  for (const rule of rules) {
    const field = rule.field as keyof RedactedMessage;
    if (!(field in result) || field === 'id') continue;

    const value = result[field];
    if (typeof value !== 'string') continue;

    if (rule.pattern) {
      result[field] = value.replace(new RegExp(rule.pattern, 'g'), '[REDACTED]');
    } else {
      result[field] = '[REDACTED]';
    }
  }

  return result as unknown as RedactedMessage;
}

File: python/steward/redactor.py
"""Deterministic, non-LLM redactor. Sits between triage and planner."""
from __future__ import annotations

import re
from typing import Any, TypedDict

from steward.rules import RedactionRule


class RedactedMessage(TypedDict, total=False):
    id: str
    fromDomain: str
    subject: str


def redact(message: dict[str, Any]) -> RedactedMessage:
    """Strip body, reduce from to domain."""
    from_addr: str = message["from"]
    at = from_addr.rfind("@")
    from_domain = from_addr[at + 1:] if at >= 0 else from_addr
    return {
        "id": message["id"],
        "fromDomain": from_domain,
        "subject": message["subject"],
    }


def apply_redaction_rules(
    message: RedactedMessage,
    rules: list[RedactionRule],
) -> RedactedMessage:
    """Apply rule-driven redaction on top of the base. Returns a new dict."""
    if not rules:
        return message
    result: dict[str, Any] = dict(message)
    for rule in rules:
        field = rule.get("field")
        if not field or field not in result or field == "id":
            continue
        value = result[field]
        if not isinstance(value, str):
            continue
        pattern = rule.get("pattern")
        if pattern:
            result[field] = re.sub(pattern, "[REDACTED]", value)
        else:
            result[field] = "[REDACTED]"
    return result  # type: ignore[return-value]
